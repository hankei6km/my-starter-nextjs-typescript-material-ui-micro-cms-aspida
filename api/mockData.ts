import { Test1Contents } from './contentTypes';

export const mockDataTest1: Test1Contents = {
  contents: [
    {
      id: 'zzzzzzzzz',
      createdAt: '2020-12-27T04:04:30.107Z',
      updatedAt: '2020-12-27T04:04:30.107Z',
      publishedAt: '2020-12-27T04:04:30.107Z',
      revisedAt: '2020-12-27T04:04:30.107Z',
      title: 'title3',
      content: '<p>content3</p>'
    },
    {
      id: 'yyyyyy-da',
      createdAt: '2020-12-26T15:29:14.476Z',
      updatedAt: '2020-12-26T15:29:14.476Z',
      publishedAt: '2020-12-26T15:29:14.476Z',
      revisedAt: '2020-12-26T15:29:14.476Z',
      title: 'title2',
      content: '<p>content2</p>'
    },
    {
      id: 'xxxxxxxxx',
      createdAt: '2020-12-26T12:25:43.532Z',
      updatedAt: '2020-12-26T12:27:22.533Z',
      publishedAt: '2020-12-26T12:27:22.533Z',
      revisedAt: '2020-12-26T12:27:22.533Z',
      title: 'title1',
      content: '<p>content1</p>'
    }
  ],
  totalCount: 3,
  offset: 0,
  limit: 10
};

export const mockDataTest1List = {
  ...mockDataTest1,
  contents: mockDataTest1.contents.map((v) => ({
    ...v,
    content: undefined
  }))
};

export const mockDataTest1Ids = {
  ...mockDataTest1,
  contents: mockDataTest1.contents.map(({ id }) => ({ id }))
};
