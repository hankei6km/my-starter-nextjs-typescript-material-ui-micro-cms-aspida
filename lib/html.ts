import unified, { Processor, FrozenProcessor } from 'unified';
import rehypeParse from 'rehype-parse';
import rehypeMinifyWhitespace from 'rehype-minify-whitespace';
import rehypeStringify from 'rehype-stringify';
import rehypeSanitize from 'rehype-sanitize';
import merge from 'deepmerge';
import gh from 'hast-util-sanitize/lib/github.json';
import { Schema } from 'hast-util-sanitize';
import cheerio from 'cheerio';
import parseStyle from 'style-to-object';
// import camelcaseKeys from 'camelcase-keys';  // vedor prefix が jsx styleにならない?
import { Section, SectionContentHtmlChildren } from '../types/pageTypes';

import visit from 'unist-util-visit';
import { Element } from 'hast';
// import { Heading } from 'mdast';
import { Transformer } from 'unified';

const schema = merge(gh, {
  tagNames: ['picture', 'source'],
  attributes: {
    source: ['srcSet', 'sizes'],
    img: ['srcSet', 'sizes', 'className']
  }
});

export function adjustHeading(
  { top }: { top: number } = { top: 4 }
): Transformer {
  // hコンテント HTML と Markdown で本文に入力される見出しは h2 と h3 .
  // h1 は保険で最上位にまるめる、.
  return function transformer(tree): void {
    // 今回は最上位が h3 か h4 に固定
    function visitorH3Top(node: Element, _index: number): void {
      switch (node.tagName) {
        case 'h1':
          node.tagName = 'h3';
          break;
        case 'h2':
          node.tagName = 'h3';
          break;
        case 'h3':
          node.tagName = 'h4';
          break;
      }
    }
    function visitorH4Top(node: Element, _index: number): void {
      switch (node.tagName) {
        case 'h1':
          node.tagName = 'h4';
          break;
        case 'h2':
          node.tagName = 'h4';
          break;
        case 'h3':
          node.tagName = 'h5';
          break;
      }
    }
    visit(tree, 'element', top === 3 ? visitorH3Top : visitorH4Top);
  };
}

export function processorHtml() {
  return unified().use(rehypeParse, { fragment: true });
}

function normalizeProcessor(processor: Processor): FrozenProcessor {
  return processor
    .use(rehypeMinifyWhitespace)
    .use(rehypeSanitize, (schema as unknown) as Schema)
    .use(rehypeStringify)
    .freeze();
}

export function normalizedHtml(processor: Processor, html: string): string {
  let ret = '';
  normalizeProcessor(processor).process(html, (err, file) => {
    if (err) {
      console.error(err);
    }
    ret = String(file);
  });
  return ret;
}

export function styleToJsxStyle(
  s: string
): SectionContentHtmlChildren['style'] {
  const p = parseStyle(s) || {};
  const ret: SectionContentHtmlChildren['style'] = {};
  Object.entries(p).forEach((kv) => {
    const [first, ...names] = kv[0].split('-');
    const capitalizedNames = names.map(
      (r) => `${r[0].toLocaleUpperCase()}${r.slice(1)}`
    );
    const k = `${first}${capitalizedNames.join('')}`;

    ret[k] = kv[1];
  });
  return ret;
}

// とりあえず、普通に記述された markdown から変換されたときに body の直下にありそうなタグ.
// いまのところ小文字のみ.
export const SectionContentHtmlChildrenElemValues = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'div',
  'p',
  'blockquote',
  'code',
  'pre',
  'span',
  'img',
  'picture',
  'table',
  'ul',
  'ol',
  'dl',
  'hr',
  'a',
  'header',
  'footer',
  'section',
  'article',
  'aside'
] as const;
export function htmlToChildren(html: string): SectionContentHtmlChildren[] {
  // markdown の変換と sanitize には unifed を使っているので、少し迷ったが cheerio を利用.
  // (unified は typescript で利用するときに対応していないプラグインがたまにあるので)

  const ret: SectionContentHtmlChildren[] = [];
  const $ = cheerio.load(html);
  try {
    $('body')
      .contents()
      .each((_idx, elm) => {
        // https://stackoverflow.com/questions/31949521/scraping-text-with-cheerio
        // if (elm.nodeType === Node.ELEMENT_NODE) {
        if (elm.nodeType === 1) {
          if (
            SectionContentHtmlChildrenElemValues.some((v) => v === elm.tagName)
          ) {
            const $elm = $(elm);
            const attribs = elm.attribs ? { ...elm.attribs } : {};
            let style: SectionContentHtmlChildren['style'] = {};
            if (elm.attribs.style) {
              style = styleToJsxStyle(elm.attribs.style);
              delete attribs.style;
            }
            if (elm.attribs.class) {
              attribs.className = elm.attribs.class;
              delete attribs.class;
            }
            ret.push({
              tagName: elm.tagName,
              style: style,
              attribs: attribs,
              html: $elm.html() || ''
            });
          } else {
            throw new Error(
              `support only ${SectionContentHtmlChildrenElemValues.join(', ')}`
            );
          }
        } else {
          throw new Error('support only nodeType=Node.ELEMENT_NODE');
        }
      });
  } catch (_e) {
    ret.splice(0, ret.length, {
      tagName: 'div',
      style: {},
      attribs: {},
      html: html
    });
  }
  if (ret.length === 0) {
    ret.push({
      tagName: 'div',
      style: {},
      attribs: {},
      html: html
    });
  }
  return ret;
}

export type IndexedHtml = {
  index: {
    range: [
      // start and end pos.
      number,
      number
    ];
    // section  section contentHtml を特定する idx.
    sectionIdx: number;
    contentIdx: number;
    childIdx: number;
  }[];
  html: string;
};

export function getIndexedHtml(sections: Section[]): IndexedHtml {
  const ret: IndexedHtml = {
    index: [],
    html: ''
  };
  sections.forEach((section, sectionIdx) => {
    section.content.forEach((content, contentIdx) => {
      if (content.kind === 'html') {
        // filter ではじくと index がずれる
        // const contentHtmlWithoutCodeBlock = content.contentHtml.filter(
        //   ({ tagName, html }) =>
        //     tagName !== 'pre' && html.slice(0, 4) !== 'code'
        // );
        content.contentHtml.forEach(({ tagName, html }, childIdx) => {
          if (
            tagName !== 'img' &&
            !(tagName === 'pre' && html.slice(0, 5) === '<code')
          ) {
            const start = ret.html.length;
            ret.html = `${ret.html}<${tagName}>${html}</${tagName}>`;
            const index: IndexedHtml['index'][0] = {
              range: [start, ret.html.length - 1],
              sectionIdx: sectionIdx,
              contentIdx: contentIdx,
              childIdx: childIdx
            };
            ret.index.push(index);
          }
        });
      }
    });
  });
  return ret;
}
