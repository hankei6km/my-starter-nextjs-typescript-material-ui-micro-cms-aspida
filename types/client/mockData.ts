import { PagesContents, ArticleContents } from './contentTypes';

export const mockDataPages: PagesContents = {
  contents: [
    {
      id: 'home',
      createdAt: '2020-12-27T04:04:30.107Z',
      updatedAt: '2020-12-27T04:04:30.107Z',
      publishedAt: '2020-12-27T04:04:30.107Z',
      revisedAt: '2020-12-27T04:04:30.107Z',
      title: 'Home',
      kind: ['page'],
      description: 'my starter home page',
      sections: [
        {
          fieldId: 'sectionContent',
          title: 'intro',
          content: [
            {
              fieldId: 'contentHtml',
              html: '<p>index page</p>'
            },
            {
              fieldId: 'contentMarkdown',
              markdown: '## markdown\ndescribed by using markdown'
            },
            {
              fieldId: 'contentMarkdown',
              markdown: '## blog'
            },
            {
              fieldId: 'contentArticles',
              apiName: 'posts'
            }
          ]
        }
      ]
    },
    {
      id: 'blog',
      createdAt: '2020-12-26T15:29:14.476Z',
      updatedAt: '2020-12-26T15:29:14.476Z',
      publishedAt: '2020-12-26T15:29:14.476Z',
      revisedAt: '2020-12-26T15:29:14.476Z',
      title: 'Blog',
      kind: ['posts'],
      sections: [
        {
          fieldId: 'sectionContent',
          title: 'test1 posts',
          content: [
            {
              fieldId: 'contentArticles',
              apiName: 'posts',
              detail: true
            }
          ]
        }
      ]
    }
  ],
  totalCount: 2,
  offset: 0,
  limit: 10
};

export const mockDataPagesList = {
  ...mockDataPages,
  contents: mockDataPages.contents.map((v) => ({
    ...v,
    kind: undefined,
    description: undefined,
    sections: undefined
  }))
};

export const mockDataPagesIds = {
  ...mockDataPages,
  contents: mockDataPages.contents.map(({ id }) => ({ id }))
};

export const mockDataArticles: ArticleContents = {
  contents: [
    {
      id: 'mmmmmmmmm',
      createdAt: '2021-01-13T05:12.157Z',
      updatedAt: '2021-01-13T05:12.157Z',
      publishedAt: '2021-01-13T05:12.157Z',
      revisedAt: '2021-01-13T05:12.157Z',
      title: 'title4',
      content: [
        {
          fieldId: 'contentMarkdown',
          markdown: 'markdown content'
        }
      ]
    },
    {
      id: 'zzzzzzzzz',
      createdAt: '2020-12-27T04:04:30.107Z',
      updatedAt: '2020-12-27T04:04:30.107Z',
      publishedAt: '2020-12-27T04:04:30.107Z',
      revisedAt: '2020-12-27T04:04:30.107Z',
      title: 'title3',
      content: [
        {
          fieldId: 'contentHtml',
          html: '<p>content3</p>'
        }
      ]
    },
    {
      id: 'yyyyyy-da',
      createdAt: '2020-12-26T15:29:14.476Z',
      updatedAt: '2020-12-26T15:29:14.476Z',
      publishedAt: '2020-12-26T15:29:14.476Z',
      revisedAt: '2020-12-26T15:29:14.476Z',
      title: 'title2',
      content: [
        {
          fieldId: 'contentHtml',
          html: '<p>content2</p>'
        }
      ]
    },
    {
      id: 'xxxxxxxxx',
      createdAt: '2020-12-26T12:25:43.532Z',
      updatedAt: '2020-12-26T12:27:22.533Z',
      publishedAt: '2020-12-26T12:27:22.533Z',
      revisedAt: '2020-12-26T12:27:22.533Z',
      title: 'title1',
      content: [
        {
          fieldId: 'contentHtml',
          html: '<p>content1</p>'
        }
      ]
    }
  ],
  totalCount: 3,
  offset: 0,
  limit: 10
};

export const mockDataArticleList = {
  ...mockDataArticles,
  contents: mockDataArticles.contents.map((v) => ({
    ...v,
    content: undefined
  }))
};

export const mockDataArticleIds = {
  ...mockDataArticles,
  contents: mockDataArticles.contents.map(({ id }) => ({ id }))
};
