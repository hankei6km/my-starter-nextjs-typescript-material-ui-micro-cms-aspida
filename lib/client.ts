import aspida from '@aspida/fetch';
import api from '../types/client/$api';
import mock from '../types/client/$mock';

export const fetchConfig = (() => {
  const getApiKey = process.env.GET_API_KEY || '';
  if (getApiKey === '' && process.env.NODE_ENV !== 'test') {
    console.error('$GET_API_KEY is not defined.');
  }
  return {
    headers: { 'X-API-KEY': getApiKey }
  };
})();

// aspida の client.xxx の xxx 部分。
// 今回は api.vi[apiName]
// 手動で定義する必要があるので注意:
export const ApiNameArticleValues = ['posts'] as const;
export type ApiNameArticle = typeof ApiNameArticleValues[number];

const clientV1 = (process.env.USE_MOCK_CLIENT_FORCE === 'true' ||
process.env.USE_MOCK_CLIENT === 'true'
  ? mock(aspida(fetch))
  : api(aspida(fetch))
).api.v1;

export default clientV1;
