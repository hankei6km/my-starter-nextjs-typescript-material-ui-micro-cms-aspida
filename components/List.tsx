import React from 'react';
import Box from '@material-ui/core/Box';
import MuiList from '@material-ui/core/List';
import MuiListItem from '@material-ui/core/ListItem';
import ListItem from './ListItem';
import ListDetail from './ListDetail';
import { PostsIndex } from '../types/client/contentTypes';

type Props = {
  items: PostsIndex[];
  detail?: boolean;
};

const List = ({ items, detail = true }: Props) => (
  <MuiList dense disablePadding>
    {items.map((item) => (
      <MuiListItem dense divider={detail} key={item.id}>
        {detail ? (
          <Box my={1} width="100%">
            <ListDetail data={item} />
          </Box>
        ) : (
          <Box my={0} width="100%">
            <ListItem data={item} />
          </Box>
        )}
      </MuiListItem>
    ))}
  </MuiList>
);

export default List;
