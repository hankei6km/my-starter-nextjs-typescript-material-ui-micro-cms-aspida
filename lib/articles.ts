import { ParsedUrlQuery } from 'querystring';
import { GetStaticPropsContext } from 'next';
import client, { fetchConfig, ApiNameArticle } from './client';
import { ArticlePost, blankArticlePost } from '../types/articleTypes';
import {
  ArticleContent,
  blankArticleContent
} from '../types/client/contentTypes';
import { markdownToHtml } from './markdown';

export async function getSortedArticleList(apiName: ApiNameArticle) {
  try {
    const res = await client[apiName].get({
      query: {
        fields: 'id,createdAt,updatedAt,publishedAt,revisedAt,title'
      },
      config: fetchConfig
    });
    return res.body.contents;
  } catch (err) {
    // res.status = 404 などでも throw される(試した限りでは)
    // res.status を知る方法は?
    console.error(`getSortedArticleList error: ${apiName}: ${err.name}`);
  }
  return [];
}

export async function getAllArticleIds(apiName: ApiNameArticle) {
  try {
    const res = await client[apiName].get({
      query: {
        fields: 'id'
      },
      config: fetchConfig
    });
    return res.body.contents.map(({ id }) => ({ params: { id } }));
  } catch (err) {
    console.error(`getAllArticleIds error: ${apiName}: ${err.name}`);
  }
  return [];
}

export async function getArticleData(
  apiName: ApiNameArticle,
  {
    params = { id: '' },
    preview = false,
    previewData = {}
  }: GetStaticPropsContext<ParsedUrlQuery>
): Promise<ArticleContent> {
  try {
    const res = await client[apiName]
      ._id(!preview ? params.id : previewData.slug) // 似たような3項式がバラけていてすっきりしない
      .$get({
        query: {
          draftKey: !preview ? '' : previewData.draftKey
        },
        config: fetchConfig
      });
    return res;
  } catch (err) {
    console.error(`getArticleData error: ${apiName}: ${err.name}`);
  }
  return blankArticleContent();
}

export async function getArticlePostData(
  apiName: ApiNameArticle,
  {
    params = { id: '' },
    preview = false,
    previewData = {}
  }: GetStaticPropsContext<ParsedUrlQuery>
): Promise<ArticlePost> {
  try {
    const res = await getArticleData(apiName, {
      params,
      preview,
      previewData
    });
    const data = {
      id: res.id,
      updated: res.revisedAt || res.updatedAt,
      title: res.title,
      content: res.content.map((content) => {
        if (content.fieldId === 'contentHtml') {
          return { kind: 'html' as 'html', html: content.html };
        } else if (content.fieldId === 'contentMarkdown') {
          return {
            kind: 'html' as 'html',
            html: markdownToHtml(content.markdown)
          };
        }
        return { kind: 'html' as 'html', html: '' };
      }),
      mainImage: res.mainImage || ''
    };
    return data;
  } catch (err) {
    console.error(`getArticlePostData error: ${apiName}: ${err.name}`);
  }
  return blankArticlePost();
}
