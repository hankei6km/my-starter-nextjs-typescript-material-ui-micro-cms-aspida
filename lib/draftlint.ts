import cheerio from 'cheerio';
import { TextlintKernel } from '@textlint/kernel';
import { TextlintKernelOptions } from '@textlint/kernel/lib/textlint-kernel-interface';
import { Section } from '../types/pageTypes';
import { getIndexedHtml } from './html';
import { getTextlintKernelOptions } from '../utils/textlint';
import traverse from 'traverse';
const parseHtml = require('textlint-plugin-html/lib/html-to-ast').parse;

type InsInfo = {
  message: string;
  ruleId: string;
  severity: number;
  index: number;
  insIndex: number;
  range: [number, number];
};
type InsInfos = InsInfo[];

export function getInsInfos(html: string, messages: any[]): InsInfos {
  const ret: InsInfos = [];
  const nodes = traverse(parseHtml(html)).nodes();
  messages.forEach((m) => {
    let hit = -1;
    const index = m.index;
    nodes.forEach((node, ii) => {
      if (
        node.type &&
        node.range &&
        node.range[0] <= index &&
        index <= node.range[1]
      ) {
        hit = ii;
      }
    });
    if (hit >= 0) {
      const node = nodes[hit];
      const end = m.index - node.range[0];
      // console.log(end, node.value.slice(0, end), node.raw.slice(0, end));
      ret.push({
        message: m.message,
        ruleId: m.ruleId,
        severity: m.severity,
        index: m.index,
        // 末尾 ';' だと欠ける? よって先頭を使う
        // TODO: 末尾の扱いチェック(どこで欠ける?).
        insIndex:
          node.value &&
          node.raw &&
          node.value.slice(0, end) === node.raw.slice(0, end)
            ? m.index
            : node.range[0] - 1,
        range: [node.range[0] as number, node.range[1] as number]
      });
    }
  });
  return ret.sort((a, b) => a.insIndex - b.insIndex);
}

export function insertHtmlToSections(
  html: string,
  pos: number, // section を IndexedHtml の html に変換した場合の html の位置
  sections: Section[]
): Section[] {
  //  const ret = [...sections];
  const { index } = getIndexedHtml(sections);
  const idx = index.findIndex(
    ({ range }) => range[0] <= pos && pos <= range[1]
  );
  if (idx >= 0) {
    const { range, sectionIdx, contentIdx, childIdx } = index[idx];
    const content = sections[sectionIdx].content[contentIdx];
    if (content.kind === 'html') {
      const c = content.contentHtml[childIdx];
      let posInChild = pos - range[0] - (c.tagName.length + 2); // 2 = '<>'.length
      posInChild = posInChild < 0 ? 0 : posInChild;
      c.html = `${c.html.slice(0, posInChild)}${html}${c.html.slice(
        posInChild
      )}`;
    }
  } // 見つからないときは?
  return sections;
}
type TextLintInSectionsResult = {
  sections: Section[];
  messages: { ruleId: string; message: string; id: string; severity: number }[];
  list: string;
};

export async function textLintInSections(
  sections: Section[],
  options?: TextlintKernelOptions,
  messageStyle: { [key: string]: string } = { color: 'red' },
  idPrefix: string = ''
): Promise<TextLintInSectionsResult> {
  let ret: TextLintInSectionsResult = { sections, messages: [], list: '' };
  const indexedHtml = getIndexedHtml(sections);

  // https://github.com/mobilusoss/textlint-browser-runner/tree/master/packages/textlint-bundler
  const kernel = new TextlintKernel();

  const results = [
    await kernel.lintText(
      indexedHtml.html,
      // todo: options(presets など)は SiteServerSideConfig で定義できるようにする.
      options || getTextlintKernelOptions()
    )
  ];
  if (results && results.length > 0 && results[0].messages.length > 0) {
    const insInfos = getInsInfos(indexedHtml.html, results[0].messages);
    let slider = 0;
    const $wrapper = cheerio.load('<span/>')('span');
    Object.entries(messageStyle).forEach(([k, v]) => {
      $wrapper.css(k, v);
    });
    insInfos.forEach((m, i) => {
      const id = `${idPrefix}:textLintMessage:${i}`;
      $wrapper.attr('id', id);
      const html = $wrapper.html(m.message).parent().html();
      if (m.message && html) {
        const index = m.insIndex + 1;
        ret.sections = insertHtmlToSections(html, index + slider, ret.sections);
        slider = slider + html.length;
        ret.messages.push({
          ruleId: m.ruleId,
          message: m.message,
          id,
          severity: m.severity
        });
      }
    });
    const $dl = cheerio.load('<dl/>')('dl');
    ret.messages.forEach((m) => {
      const $d = cheerio.load('<dt></dt><dd><a/></dd>');
      // https://github.com/textlint/textlint/blob/master/packages/%40textlint/kernel/src/shared/rule-severity.ts
      // https://github.com/textlint/textlint/blob/master/packages/%40textlint/kernel/src/context/TextlintRuleSeverityLevelKeys.ts
      $d('dt').text(
        m.severity === 0 ? 'info' : m.severity === 1 ? 'warning' : 'error'
      );
      const $a = $d('a');
      $a.attr('href', `#${m.id}`);
      $a.text(`${m.message}(${m.ruleId})`);
      $dl.append($d('body').children());
    });
    ret.list = $dl.parent().html() || '';
  }
  return ret;
}
