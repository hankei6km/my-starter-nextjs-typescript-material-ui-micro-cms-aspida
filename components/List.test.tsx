import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { RouterContext } from 'next/dist/next-server/lib/router-context';
import { mockRouter, mockImage } from '../test/testUtils';

import { mockDataTest1List } from '../types/client/mockData';
import List from './List';

// https://stackoverflow.com/questions/40449434/mocking-globals-in-jest
const orgImage = global.Image;
afterEach(() => {
  global.Image = orgImage;
});

describe('List', () => {
  it('should renders List', async () => {
    const addEventListener = jest.fn();
    global.Image = mockImage(jest.fn(), addEventListener);

    await act(async () => {
      const router = mockRouter();
      const { container } = render(
        <RouterContext.Provider value={router}>
          <List items={mockDataTest1List.contents} />
        </RouterContext.Provider>
      );

      await new Promise((resolve) => {
        setTimeout(() => {
          addEventListener.mock.calls.forEach((v) => {
            v[1]();
          });
          resolve();
        }, 10);
      });
      expect(container).toMatchSnapshot();
    });
  });
});
