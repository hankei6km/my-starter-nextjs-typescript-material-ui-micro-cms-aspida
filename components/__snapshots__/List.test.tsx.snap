// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should renders List 1`] = `
<div>
  <ul
    class="MuiList-root MuiList-dense"
  >
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-divider"
    >
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-content-3 MuiPaper-elevation0 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiCardHeader-root makeStyles-title-2"
            >
              <div
                class="MuiCardHeader-content"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-root MuiTypography-h4 MuiTypography-colorTextPrimary MuiTypography-colorPrimary"
                  href="/test1/zzzzzzzzz"
                >
                  title3
                </a>
                <span
                  class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  <div
                    class="MuiBox-root MuiBox-root-7 makeStyles-root-6"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                      <path
                        d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                      />
                    </svg>
                    <div
                      class="MuiTypography-root MuiTypography-body2"
                    >
                      <time
                        datetime="2020-12-27T04:04:30.107Z"
                      >
                        2020-12-27
                      </time>
                    </div>
                  </div>
                </span>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-8 makeStyles-thumb-4"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                href="/test1/zzzzzzzzz"
              >
                <div
                  class="makeStyles-thumb-9"
                >
                  <img
                    alt="thumbnail for title3"
                    height="150"
                    width="250"
                  />
                </div>
              </a>
            </div>
          </div>
          <div
            class="MuiCardActions-root makeStyles-actions-5 MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
              href="/test1/zzzzzzzzz"
              role="button"
              style="text-transform: none;"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                read more
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-divider"
    >
      <div
        class="MuiBox-root MuiBox-root-10"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-content-3 MuiPaper-elevation0 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiCardHeader-root makeStyles-title-2"
            >
              <div
                class="MuiCardHeader-content"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-root MuiTypography-h4 MuiTypography-colorTextPrimary MuiTypography-colorPrimary"
                  href="/test1/yyyyyy-da"
                >
                  title2
                </a>
                <span
                  class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  <div
                    class="MuiBox-root MuiBox-root-11 makeStyles-root-6"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                      <path
                        d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                      />
                    </svg>
                    <div
                      class="MuiTypography-root MuiTypography-body2"
                    >
                      <time
                        datetime="2020-12-26T15:29:14.476Z"
                      >
                        2020-12-26
                      </time>
                    </div>
                  </div>
                </span>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-12 makeStyles-thumb-4"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                href="/test1/yyyyyy-da"
              >
                <div
                  class="makeStyles-thumb-9"
                >
                  <img
                    alt="thumbnail for title2"
                    height="150"
                    width="250"
                  />
                </div>
              </a>
            </div>
          </div>
          <div
            class="MuiCardActions-root makeStyles-actions-5 MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
              href="/test1/yyyyyy-da"
              role="button"
              style="text-transform: none;"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                read more
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-divider"
    >
      <div
        class="MuiBox-root MuiBox-root-13"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-content-3 MuiPaper-elevation0 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiCardHeader-root makeStyles-title-2"
            >
              <div
                class="MuiCardHeader-content"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-root MuiTypography-h4 MuiTypography-colorTextPrimary MuiTypography-colorPrimary"
                  href="/test1/xxxxxxxxx"
                >
                  title1
                </a>
                <span
                  class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  <div
                    class="MuiBox-root MuiBox-root-14 makeStyles-root-6"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                      <path
                        d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                      />
                    </svg>
                    <div
                      class="MuiTypography-root MuiTypography-body2"
                    >
                      <time
                        datetime="2020-12-26T12:27:22.533Z"
                      >
                        2020-12-26
                      </time>
                    </div>
                  </div>
                </span>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-15 makeStyles-thumb-4"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                href="/test1/xxxxxxxxx"
              >
                <div
                  class="makeStyles-thumb-9"
                >
                  <img
                    alt="thumbnail for title1"
                    height="150"
                    width="250"
                  />
                </div>
              </a>
            </div>
          </div>
          <div
            class="MuiCardActions-root makeStyles-actions-5 MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
              href="/test1/xxxxxxxxx"
              role="button"
              style="text-transform: none;"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                read more
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
